### ISSUE TITLE
Server crash when trying to access menu Jobs

### ISSUE THREAD

**Issue by vgoineau** [opencv/cvat#5567]
CVAT is install on Docker Swarm has a stack. All component works fine on startup. We added two projects with multiple tasks and each task has multiple jobs. More than 100 jobs. As soon as one user click on Jobs menu (top menu). Server backend crash, and nothing is working. It gives 504 error and get backend crash. 

### Expected Behaviour
When multiple user does multiple request, server should be able give response. Looks like it is backend limits the number of concurrent sessions. We might need to increase the number of session on http server build in docker image. 

### Steps to Reproduce (for bugs)
1. Fresh install cvat on docker swarm
2. add projects with multiple tasks and as many jobs has possible
3. click on jobs in top menu and see what happen
4. You should see something like : 

![image](https://user-images.githubusercontent.com/13660594/211648024-7e42c650-c8c9-4461-840e-e2f9394773f1.png)

### Context
Just want to search and filter jobs inside CVAT

### Your Environment
- Docker version 20.10.21
- Swarm setup with 1 master and 5 worker
- OS is on Linux Ubuntu 22.04

- Logs cvat_server :
`2023-01-10 19:47:56,425 DEBG 'runserver' stderr output:
[Tue Jan 10 19:47:56.423106 2023] [wsgi:info] [pid 581:tid 139722279259200] mod_wsgi (pid=1649): Process 'localhost:8080' has died, deregister and restart it.
[Tue Jan 10 19:47:56.423169 2023] [wsgi:info] [pid 581:tid 139722279259200] mod_wsgi (pid=1649): Process 'localhost:8080' has been deregistered and will no longer be monitored.
[Tue Jan 10 19:47:56.424178 2023] [wsgi:info] [pid 1677:tid 139722279259200] mod_wsgi (pid=1677): Starting process 'localhost:8080' with threads=5.
[Tue Jan 10 19:47:56.426035 2023] [wsgi:info] [pid 1677:tid 139722279259200] mod_wsgi (pid=1677): Python home /opt/venv.
[Tue Jan 10 19:47:56.426210 2023] [wsgi:info] [pid 1677:tid 139722279259200] mod_wsgi (pid=1677): Initializing Python.
2023-01-10 19:47:56,426 DEBG 'runserver' stderr output:
2023-01-10 19:47:56,466 DEBG 'runserver' stderr output:
[Tue Jan 10 19:47:56.465932 2023] [wsgi:info] [pid 1677:tid 139722279259200] mod_wsgi (pid=1677): Attach interpreter ''.
2023-01-10 19:47:56,470 DEBG 'runserver' stderr output:
[Tue Jan 10 19:47:56.469731 2023] [wsgi:info] [pid 1677:tid 139722279259200] mod_wsgi (pid=1677): Imported 'mod_wsgi'.
[Tue Jan 10 19:47:56.469823 2023] [wsgi:info] [pid 1677:tid 139722279259200] mod_wsgi (pid=1677, process='localhost:8080', application=''): Loading Python script file '/tmp/cvat-server/handler.wsgi'.
2023-01-10 19:47:57,833 DEBG 'runserver' stderr output:
[Tue Jan 10 19:47:57.833366 2023] [wsgi:error] [pid 1677:tid 139721986103040] [remote 10.0.2.57:57510] mod_wsgi (pid=1677): Queue timeout expired for WSGI daemon process 'localhost:8080'.
2023-01-10 19:47:57,834 DEBG 'runserver' stderr output:
[Tue Jan 10 19:47:57.833976 2023] [wsgi:error] [pid 1677:tid 139721737668352] [remote 10.0.2.57:57520] mod_wsgi (pid=1677): Queue timeout expired for WSGI daemon process 'localhost:8080'.
[Tue Jan 10 19:47:57.834242 2023] [wsgi:error] [pid 1677:tid 139721729275648] [remote 10.0.2.57:57526] mod_wsgi (pid=1677): Queue timeout expired for WSGI daemon process 'localhost:8080'.
[Tue Jan 10 19:47:57.834429 2023] [wsgi:error] [pid 1677:tid 139721746061056] [remote 10.0.25.10:44976] mod_wsgi (pid=1677): Queue timeout expired for WSGI daemon process 'localhost:8080'.
2023-01-10 19:47:57,835 DEBG 'runserver' stderr output:
[Tue Jan 10 19:47:57.834880 2023] [wsgi:error] [pid 1677:tid 139721720882944] [remote 10.0.25.10:51278] mod_wsgi (pid=1677): Queue timeout expired for WSGI daemon process 'localhost:8080'.
[Tue Jan 10 19:47:52.431243 2023] [wsgi:error] [pid 673:tid 139722271422208] [client 10.0.2.57:57526] Timeout when reading response headers from daemon process 'localhost:8080': /tmp/cvat-server/htdocs/api, referer: https://cvat.domain.com/jobs?page=2
2023-01-10 19:47:52,948 DEBG 'runserver' stderr output:
[Tue Jan 10 19:47:52.948473 2023] [wsgi:error] [pid 640:tid 139722266134272] [client 10.0.25.10:44976] Timeout when reading response headers from daemon process 'localhost:8080': /tmp/cvat-server/htdocs/api
2023-01-10 19:47:55,752 DEBG 'runserver' stderr output:
[Tue Jan 10 19:47:55.751985 2023] [wsgi:info] [pid 1649:tid 139721585977088] mod_wsgi (pid=1649): Aborting process 'localhost:8080'.
[Tue Jan 10 19:47:55.752038 2023] [wsgi:info] [pid 1649:tid 139721585977088] mod_wsgi (pid=1649): Exiting process 'localhost:8080'.
2023-01-10 19:47:56,424 DEBG 'runserver' stderr output:`

- Console Logs : 
![image](https://user-images.githubusercontent.com/13660594/211645697-0eadebc6-7b48-4c03-9661-75d377ad29ed.png)

**Comment by bsekachev** [comment]
@azhavoro 

Do you have any ideas?

**Comment by vgoineau** [comment]
@azhavoro : well I’m a bit stuck to be honest. 

### My hypothesis is

- cvat_ui (front end) does multiple requests to the cvat_server at once per user. 
- backend receive theses request but because it takes more than average to process. Concurrent request get full. 
- request on V2.3 is manage by Apache for some reason (I’m not sure of that but it looks like that)
- Apache config when container start is very low. Meaning that Maxworker request setting for example is at 5 or something. 
- Apache crashes and so all the django container can work. 

To see if this hypothesis was right here three tests that I did.

### Test 1 : change Apache settings

To test this hypothesis I tried to change Apache config files but nothing happen server side. Meaning that even if configs are changed Apache still keep initial setting. I did not try to rebuild docker image. I still use the opencv/Cvat provided on docker hub. 
 
For example one Apache setting that I tried to change is : MaxRequestWorkers 

### Test 2 : scaling backend services
Also, another possible solution would have to scale the replicas of cvat_server inside docker swarm. This would have an effect of load balancing all request across multiple replicas of cvat_server. And so increasing the chance of not having a total crash. 

But because django cvat_server is using data caching that use SQLite… well we have a DB locking problem which means that SQLite have concurrent connexion to same database… and gets the service to crash. 

### Test 3 : increasing NUMPROCS 

third and last test a try is increase numprocs inside my docker-compose. This config is a env variable for docker to deploy my stack. So by increasing that to 3 or 5… service does not start because process name is not define. 

And I do not know how to define this variable so supervisord can start according to numproc > 1. 


So this is where I am now. And I would appreciate any sort of help. Like I said… Im a bit stuck.

**Comment by vgoineau** [comment]
@azhavoro 
Which volume needs to be local vs which one could be nfs ? Do you have info on that ?

thanks again for your time

**Comment by azhavoro** [comment]
Currently there is no any dedicated volume for cache, but you can mount a local volumes at `/home/django/data/cache` path. Other volumes could be nfs

**Comment by vgoineau** [comment]
@azhavoro 
So thanks for your time ! Putting cvat_data has local volume fixed all my problems. :+1:  Which are : 

- The crash that this ticket is describing
- But also slowlyness that we were experiencing. Loading of CVAT was slow and something we had 502 or gateway timeout error. 
- Deployment of services inside our swarm was failing

Just for documentation purposes :

For docker swarm, volume cvat_data has to be on local machine. I did sync between local stoarge and share storage, so CVAT could be deploy between host. All other volumes are nfs.

I did not try to replicas cvat_ui or cvar_server... so this has to be tested. 

Also, we  changed our OS. We are now using  ubuntu18.04 for CVAT instead of 22.04. It seams to be more stable and deployment goes smoothly without services crashing. 

So... Voilà ! Thanks again
